### PR TITLE
Multi AZ failure domain support

### PR DESCRIPTION
**Pull request by srm09**
**What this PR does / why we need it**:
This PR introduced support for Failure Domains in v1alpha4 API.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
Adds support for failure domains
```

### CODE REVIEW

File: api/v1alpha4/vspheredeploymentzone_webhook.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha4-vspheredeploymentzone,mutating=true,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=vspheredeploymentzones,verbs=create;update,versions=v1alpha4,name=validation.vspheredeploymentzone.infrastructure.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+
+var _ webhook.Defaulter = &VSphereDeploymentZone{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *VSphereDeploymentZone) Default() {
+	if r.Spec.ControlPlane == nil {
```
Comment (gab-satchi) [inline api/v1alpha4/vspheredeploymentzone_webhook.go:37]: is this different from setting `kubebuilder:default`

File: api/v1alpha4/vspherefailuredomain_webhook.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+//+kubebuilder:webhook:path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha4-vspherefailuredomain,mutating=true,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=vspherefailuredomains,verbs=create;update,versions=v1alpha4,name=default.vspherefailuredomain.infrastructure.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+
+var _ webhook.Validator = &VSphereFailureDomain{}
+var _ webhook.Defaulter = &VSphereFailureDomain{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *VSphereFailureDomain) ValidateCreate() error {
+	var allErrs field.ErrorList
+
+	if r.Spec.Topology.Datacenter == "" {
```
Comment (gab-satchi) [inline api/v1alpha4/vspherefailuredomain_webhook.go:46]: can also be moved to an OpenAPI validation

File: api/v1alpha4/vspheredeploymentzone_webhook.go (hunk 1 above)
Comment (srm09) [reply api/v1alpha4/vspheredeploymentzone_webhook.go:37]: Not sure if this can be done via the annotation since this is a pointer, can try this out.

**Review by srm09 — COMMENTED** [review]
internal reviews 1 & 2

**Comment by yastij** [comment]
@srm09 - do you want to update this or cherry-pick the current v1alpha3 implementation ?

**Comment by srm09** [comment]
> @srm09 - do you want to update this or cherry-pick the current v1alpha3 implementation?
Done, updated the PR with the cherry-picked commit from the release-0.7 branch.

**Review by yastij — APPROVED** [review]
/approve
/lgtm

**Pull request merged**